src/Build/Evaluation/Expander.cs

```diff
@@ -5493,13 +5502,57 @@ private object LateBindExecute(Exception ex, BindingFlags bindingFlags, object o
     /// <summary>
     /// This class wraps information about properties which have been used before they are initialized.
     /// </summary>
-    internal sealed class UsedUninitializedProperties
+    internal sealed class PropertiesUsageTracker
     {
+        internal LoggingContext LoggingContext { get; init; }
+
+        public PropertiesUsageTracker(LoggingContext loggingContext) => LoggingContext = loggingContext;
+
+        /// <summary>
+        /// Whether to warn when we set a property for the first time, after it was previously used.
+        /// Default is false, unless MSBUILDWARNONUNINITIALIZEDPROPERTY is set.
+        /// </summary>
+        // This setting may change after the build has started, therefore if the user has not set the property to true on the build parameters we need to check to see if it is set to true on the environment variable.
+        private bool _warnForUninitializedProperties = BuildParameters.WarnOnUninitializedProperty || Traits.Instance.EscapeHatches.WarnOnUninitializedProperty;
+
         /// <summary>
         /// Lazily allocated collection of properties and the element which used them.
         /// </summary>
         private Dictionary<string, IElementLocation>? _properties;
 
+        internal void TrackRead(string propertyName, int startIndex, int endIndex, IElementLocation elementLocation, bool isUninitialized, bool isArtificial)
+        {
+            if (isArtificial)
+            {
+                return;
+            }
+
+            if (!isUninitialized)
+            {
+                return;
+            }
```

Comment:
nit: This feels slightly unusual. Why not make the call only `if (isUninitialized && !isArtificial)` ?